src/Model/filter.ts

```diff
@@ -0,0 +1,398 @@
+var util = require('../util');
+import { Model } from './Model';
+import { type Segments } from './types';
+import * as defaultFns from './defaultFns';
+import type { Path, PathLike } from '../types';
+
+interface PaginationOptions {
+  skip?: number;
+  limit?: number;
+}
+
+type FilterFn<S> =
+  | ((item: S, key: string, object: { [key: string]: S }) => boolean)
+  | null;
+type SortFn<S> = (a: S, B: S) => number;
+
+declare module './Model' {
+  interface Model {
+    /**
+     * Creates a live-updating list from items in an object, which results in
+     * automatically updating as the input items change.
+     *
+     * @param inputPath - Path pointing to an object or array. The path's value is
+     *   retrieved via model.get(), and each item checked against filter function
+     * @param additionalInputPaths - Other parameters can be set in the model, and
+     *   the filter function will be re-evaluated when these parameters change as well.
+     * @param options
+     *   skip - The number of first results to skip
+     *   limit - The maximum number of results. A limit of zero is equivalent to no limit.
+     * @param fn - A function or the name of a function defined via model.fn(). The function
+     *   should have the arguments function(item, key, object, additionalInputs...)
+     *
+     * @see https://derbyjs.com/docs/derby-0.10/models/filters-and-sorts
+     */
+    filter<S>(
+      inputPath: PathLike,
+      additionalInputPaths: PathLike[],
+      options: PaginationOptions,
+      fn?: FilterFn<S>
+    ): Filter<S>;
+    filter<S>(
+      inputPath: PathLike,
+      additionalInputPaths: PathLike[],
+      fn?: FilterFn<S>
+    ): Filter<S>;
+    filter<S>(
+      inputPath: PathLike,
+      options: PaginationOptions,
+      fn?: FilterFn<S>
+    ): Filter<S>;
+    filter<S>(
+      inputPath: PathLike,
+      fn?: FilterFn<S>
+    ): Filter<S>;
```

Comment:
one TS-unsupported use case is having the last param be a string for a named fn. Supported in code.